apps/tools/blocks-crawler/src/blocks.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+use mina_rs_base::types::ExternalTransition;
+use mina_rs_base::JsonSerializationType;
+use serde::{Deserialize, Serialize};
+use serde_json::Error;
+
+// To read query response from graphql api (https://graphql.minaexplorer.com/)
+#[derive(Debug, Serialize, Deserialize)]
+pub struct QueryResponse {
+    pub data: Blocks,
+}
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Blocks {
+    pub blocks: Vec<BlockInfo>,
+}
+#[derive(Debug, Serialize, Deserialize)]
+pub struct BlockInfo {
+    #[serde(rename = "blockHeight")]
+    pub block_height: usize,
+    #[serde(rename = "stateHash")]
+    pub state_hash: String,
+}
+
+impl std::fmt::Display for BlockInfo {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}-{}", self.block_height, self.state_hash)
+    }
+}
+
+pub async fn check_block(block: BlockInfo) {
+    info!("Checking: {}", block);
+    // Retrieve block json from mainnet json dump
+    match retrieve_block_json("mainnet", &block).await {
```

Comment:
What if the block has already been processed? It should avoid downloading the same block again and again by using the output as a cache

Reply:
fixed